File: SMEOihaneSaez/R/plot.R
#' Funcion para visualizar la curva ROC
#'
#' @description Esta funcion calcula la Tasa de Falsos positivos (FPR) y la Tasa de Verdaderos Positivos (TPR)
#' para cada valor de corte y los grafica junto con la curva del random chance
#' @param x será un data frame con una columna lógica o binaria y las demás numéricas
#' @details Utiliza la funcion \code{\link{AUC}} dentro de la funcion para la visualizacion [mira la documentación correspondiente]
#' @return Un gráfico de la curva ROC
#' @examples
#' set.seed(123)
#' data_frame <- data.frame(
#' Variable1 = rnorm(100),  # 100 números aleatorios
#' VariableBinaria = factor(sample(c(0, 1), 100, replace = TRUE)))
#' plot_AUC(data_frame)

plot_AUC <- function(x) { #x será una matriz/ data frame
  data_AUC <- AUC(x)
  plot(data_AUC$result, type="l", col="blue", main="La curva ROC", xlab="Tasa de Falsos Positivos (FPR)",
       ylab="Tasa de Verdaderos Positivos (TPR)", xlim=c(0, 1), ylim=c(0,1))
  abline(a=0, b=1, col="red", lty=2) #la linea que define la probabilidad del 50% (es decir random)
  legend("bottomright",               # Especificar la posición de la leyenda
         legend = c("Curva ROC", "Random Chance"), # Textos de la leyenda
         col = c("blue", "red"),      # Colores de las líneas
         lty = c(1, 2),                # Estilos de las líneas (1 para sólida, 2 para punteada)
         lwd = 2)                     # Ancho de las líneas

}

#' Funcion para visualizar la entropia normalizada de las variables de un data set
#'
#' @description Esta funcion calcula la entropia normalizada para luego poder hacer un grafico de barras con cada variable
#' @param df será un data frame con variables categóricas de tipo factor
#' @details Utiliza la funcion \code{\link{norm_entropy_df}} dentro de la funcion para la visualizacion [mira la documentación correspondiente]
#' @return Un gráfico de barras para la entropia normalizada de cada variable
#' @examples
#' data <- data.frame(Columna_A = as.factor(c("Rojo", "Rojo", "Azul", "Verde", "Rojo", "Azul", "Verde", "Verde")),
#' Columna_B = as.factor(c("Alto", "Medio", "Medio", "Bajo", "Alto", "Bajo", "Alto", "Medio")),
#' Columna_C = as.factor(c("Bajo", "Alto", "Medio", "Bajo", "Medio", "Bajo", "Alto", "Medio")))
#' plot_entropy(data)

plot_entropy <- function(df) { #data frame de categorias
  H_norm <- norm_entropy_df(df)
  n<- length(H_norm)
  names_bar <- seq(n)
  barplot(H_norm, xlab="Número de variables", ylab="Entropía normalizada",
          main="Diagrama de barras para la entropía normalizada de cada variable",
          space=0, col="orange", las=1, names.arg=names_bar)

}

#' Funcion para visualizar el heatmap de una matriz de correlaciones
#'
#' @description Esta funcion calcula la matriz de correlaciones para luego poder hacer un mapa de calor
#' @param x será un data frame / matriz con variables numéricas
#' @details Utiliza la funcion \code{\link{cor_fun}} dentro de la funcion para la visualizacion [mira la documentación correspondiente]
#' @return Un mapa de calor para la matriz de correlaciones
#' @examples
#' set.seed(123)
#' datos_ejemplo <- data.frame(
#' Var1 = rnorm(100),
#' Var2 = rnorm(100),
#'  Var3 = rnorm(100),
#'  Var4 = rnorm(100))
#'  plot_cor(datos_ejemplo)
plot_cor <- function(x) { #x es un conjunto de datos matriz / data frame
  correlaciones <- cor_fun(x)
  nombres_columnas <- colnames(x)
  heatmap(correlaciones, main="Mapa de calor para la matriz de correlaciones",
          Rowv = NA, Colv = NA, labRow = nombres_columnas, labCol = nombres_columnas )
}


#' Funcion para visualizar la informacion mutua de un data set
#'
#' @description Esta funcion calcula la matriz de la información mutua para luego poder hacer un mapa de calor
#' @param x será un data frame / matriz con variables categóricas de tipo factor
#' @details Utiliza la funcion \code{\link{info_mutua}} dentro de la funcion para la visualizacion [mira la documentación correspondiente]
#' @return Un mapa de calor para la matriz de informacion mutua
#' @examples
#' data <- data.frame(Columna_A = as.factor(c("Rojo", "Rojo", "Azul", "Verde", "Rojo", "Azul", "Verde", "Verde")),
#' Columna_B = as.factor(c("Alto", "Medio", "Medio", "Bajo", "Alto", "Bajo", "Alto", "Medio")),
#' Columna_C = as.factor(c("Bajo", "Alto", "Medio", "Bajo", "Medio", "Bajo", "Alto", "Medio")))
#' plot_infomutua(data)
plot_infomutua <- function(x) {#x es un conjunto de datos categoricos
  matriz_im <- info_mutua(x)
  nombres_columnas <- colnames(x)
  heatmap(matriz_im, main="Mapa de calor para la matriz de información mutua",
          Rowv = NA, Colv = NA, labRow = nombres_columnas, labCol = nombres_columnas )
}

File: SMEOihaneSaez_py/SMEOihaneSaez_py/plot.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import AUC
from AUC import AUC
from entropia import norm_entropy_df, info_mutua
from fun_analisis import cor_fun

def plot_AUC(df):
    """
    Graficar la curva ROC utilizando la función AUC definida en el módulo AUC.
    
    Args:
        df: DataFrame que debe contener una columna numérica y una columna binaria.
        
    Returns: 
        un gráfico de la curva ROC
        
    Example:
        # Establecer semilla para reproducibilidad
        np.random.seed(123)

        # Crear un DataFrame con una variable numérica y una variable binaria
        data_frame = pd.DataFrame({
            'Variable1': np.random.normal(size=100),  # 100 números aleatorios
            'VariableBinaria': pd.Categorical(np.random.choice([0, 1], size=100))  # Variable binaria
        })

        # Graficar la curva ROC
        plot_AUC(data_frame)
    """
    # Asegúrate de que la columna binaria sea un factor
    if df[df.columns[1]].dtype.name != 'category':
        df[df.columns[1]] = df[df.columns[1]].astype('category')

    # Obtener los resultados del AUC
    data_AUC = AUC(df)
    
    # Graficar la curva ROC
    plt.figure(figsize=(8, 6))
    plt.plot(data_AUC['result'][:, 0], data_AUC['result'][:, 1], color='blue', label='Curva ROC')
    plt.plot([0, 1], [0, 1], color='red', linestyle='--', label='Probabilidad al azar (50%)')
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.0])
    plt.xlabel('Tasa de Falsos Positivos (FPR)')
    plt.ylabel('Tasa de Verdaderos Positivos (TPR)')
    plt.title('La curva ROC')
    plt.legend(loc='lower right')
    plt.grid()
    plt.show()


def plot_entropy(df):
    """
    Genera un diagrama de barras para la entropía normalizada de cada variable en un DataFrame de categorías.
    
    Args: 
        df (pd.DataFrame): un data frame de tipo categórico
        
    Returns:
        Un gráfico de barras para la entropía normalizada.
        
    Example:
        data = pd.DataFrame({
            'Columna_A': ["Rojo", "Rojo", "Azul", "Verde", "Rojo", "Azul", "Verde", "Verde"],
            'Columna_B': ["Círculo", "Cuadrado", "Círculo", "Cuadrado", "Cuadrado", "Círculo", "Triángulo", "Círculo"],
            'Columna_C': ["Bajo", "Alto", "Medio", "Bajo", "Medio", "Bajo", "Alto", "Medio"]
        })

        # Asegurarse de que las columnas son de tipo 'category'
        for col in data.columns:
            data[col] = data[col].astype('category')

        # Llamar a la función plot_entropy para graficar la entropía normalizada de cada columna
        plot_entropy(data)
    """
    
    # Calcula la entropía normalizada de cada columna en el DataFrame
    H_norm = norm_entropy_df(df)  # Asegúrate de tener una función `norm_entropy_df` que calcule la entropía normalizada
    
    # Número de variables
    n = len(H_norm)
    names_bar = range(1, n + 1)  # Genera etiquetas para cada variable
    
    # Crear el diagrama de barras
    plt.figure(figsize=(10, 6))
    plt.bar(names_bar, H_norm, color="orange")
    plt.xlabel("Número de variables")
    plt.ylabel("Entropía normalizada")
    plt.title("Diagrama de barras para la entropía normalizada de cada variable")
    plt.xticks(names_bar)  # Coloca los números de variable en el eje x
    plt.show()


def plot_cor(x):
    """
    calcula la matriz de correlaciones para luego poder hacer un mapa de calor
    
    Args:
        x (pd.DataFrame): un data frame de datos numéricos
        
    Returns:
        Un mapa de calor para la matriz de correlaciones
        
    Example:
        np.random.seed(123)

        # Crear un DataFrame de ejemplo con 100 filas y 4 variables aleatorias
        datos_ejemplo = pd.DataFrame({
            'Var1': np.random.normal(size=100),
            'Var2': np.random.normal(size=100),
            'Var3': np.random.normal(size=100),
            'Var4': np.random.normal(size=100)
        })

        plot_cor(datos_ejemplo)
    """
    # Calcular la matriz de correlaciones
    correlaciones = cor_fun(x)
    
    # Crear el mapa de calor
    plt.figure(figsize=(10, 8))
    plt.imshow(correlaciones, cmap='coolwarm', interpolation='nearest')
    plt.colorbar()  # Muestra la barra de color al lado
    
    # Configurar etiquetas usando los nombres del DataFrame original
    plt.xticks(ticks=np.arange(len(x.columns)), labels=x.columns, rotation=45)
    plt.yticks(ticks=np.arange(len(x.columns)), labels=x.columns)

    # Añadir título
    plt.title("Mapa de calor para la matriz de correlaciones")
    
    # Mostrar el gráfico
    plt.tight_layout()  # Ajustar el layout para que no se superpongan las etiquetas
    plt.show()
    


def plot_infomutua(x):
    """
    Funcion para visualizar la informacion mutua de un data set
    
    Args:
        x (pd.DataFrame): un data frame de datos categóricos
        
    Returns:
        Un mapa de calor para la matriz de informacion mutua
        
    Example:
        data = pd.DataFrame({
            'Columna_A': pd.Categorical(["Rojo", "Rojo", "Azul", "Verde", "Rojo", "Azul", "Verde", "Verde"]),
            'Columna_B': pd.Categorical(["Alto", "Medio", "Medio", "Bajo", "Alto", "Bajo", "Alto", "Medio"]),
            'Columna_C': pd.Categorical(["Bajo", "Alto", "Medio", "Bajo", "Medio", "Bajo", "Alto", "Medio"])
        })

        # Llamar a la función para graficar
        plot_infomutua(data)

    """
    # Calcular la matriz de información mutua
    matriz_im = info_mutua(x)
    
    # Crear el mapa de calor
    plt.figure(figsize=(10, 8))
    plt.imshow(matriz_im, cmap='coolwarm', interpolation='nearest')
    
    # Mostrar la barra de color al lado
    plt.colorbar()
    
    # Obtener los nombres de las columnas
    nombres_columnas = x.columns
    
    # Configurar etiquetas usando los nombres del DataFrame original
    plt.xticks(ticks=np.arange(len(nombres_columnas)), labels=nombres_columnas, rotation=45)
    plt.yticks(ticks=np.arange(len(nombres_columnas)), labels=nombres_columnas)
    
    # Añadir título
    plt.title("Mapa de calor para la matriz de información mutua")
    
    # Mostrar el gráfico
    plt.tight_layout()  # Ajustar el layout para que no se superpongan las etiquetas
    plt.show()


